### PR TITLE
Add support for running on macOS with the target jsc build

### PR DESCRIPTION
**Pull request by Eziodeng**
Cheak whether this cannot run on macOS and restore the fuzzbuild shell for macOS build which use the build-jsc shell without not officially supported "--jsc-only" port

Also fix the REPRL and main fuzzer code to deal with the jsc executable path input which lead to the executeable file cannot run on macOS for the env not properly set

The problem is that on macOS with the test tag I have used is "releases/Apple/Safari-14.0.3-macOS-11.2.3", and on the ubuntu it has coverage edges almost 960000+ and only 12000+ on mac.（Here I have found that this problem is caused the JavascriptCore.framework build, I now try to change it like WTF to build a .a library）

### CODE REVIEW

**Comment by Eziodeng** [comment]
Maybe I should first try to solve the static library of JavascriptCore.
Close this PR.

**Comment by Eziodeng** [comment]
Successfully build a static library version with some mentally settings.
Now try running for a test
Reopen it.

File: Sources/Fuzzilli/Execution/REPRL.swift [hunk 1]
```diff
@@ -49,6 +49,14 @@ public class REPRL: ComponentBase, ScriptRunner {
         self.maxExecsBeforeRespawn = maxExecsBeforeRespawn
         super.init(name: "REPRL")
 
+        #if os(macOS)
```
Comment (saelo) [inline Sources/Fuzzilli/Execution/REPRL.swift:52]: This is a bit of an unfortunate workaround. If there's no way to avoid this with JSC build options, then I think it should go into the [JSCProfile](https://github.com/googleprojectzero/fuzzilli/blob/main/Sources/FuzzilliCli/Profiles/JSCProfile.swift)

File: Sources/Fuzzilli/Modules/NetworkSync.swift [hunk 2]
```diff
@@ -48,7 +48,7 @@ fileprivate protocol MessageHandler {
 /// A connection to a network peer that speaks the above protocol.
 fileprivate class Connection {
     /// The file descriptor on POSIX or SOCKET handle on Windows of the socket.
-    let socket: libsocket.socket_t
+    let socket: libsocket.socket_fd
```
Comment (saelo) [inline Sources/Fuzzilli/Modules/NetworkSync.swift:51]: Not necessarily opposed to this rename, but why is it necessary?

File: Sources/FuzzilliCli/Profiles/JSCProfile.swift [hunk 3]
```diff
@@ -49,8 +49,12 @@ let jscProfile = Profile(
             "--thresholdForFTLOptimizeAfterWarmUp=1000",
             "--thresholdForFTLOptimizeSoon=1000",
             // Enable bounds check elimination validation
-            "--validateBCE=true",
+            //"--validateBCE=true",
```
Comment (saelo) [inline Sources/FuzzilliCli/Profiles/JSCProfile.swift:52]: What's the reason for this change? Is that feature only available on linux? I'm not sure how important it even is, maybe we can drop it entirely

File: Targets/JavaScriptCore/fuzzbuild.sh [hunk 4]
```diff
@@ -18,6 +18,10 @@ export WEBKIT_OUTPUTDIR=FuzzBuild
 
 if [ "$(uname)" == "Linux" ]; then
     ./Tools/Scripts/build-jsc --jsc-only --debug --cmakeargs="-DENABLE_STATIC_JSC=ON -DCMAKE_C_COMPILER='/usr/bin/clang' -DCMAKE_CXX_COMPILER='/usr/bin/clang++' -DCMAKE_CXX_FLAGS='-fsanitize-coverage=trace-pc-guard -O3 -lrt'"
+elif [ "$(uname)" == "Darwin" ]; then
```
Comment (saelo) [inline Targets/JavaScriptCore/fuzzbuild.sh:21]: While you're here, is our [jsc patch](https://github.com/googleprojectzero/fuzzilli/tree/main/Targets/JavaScriptCore/Patches) still necessary? I believe JSC now has native support for Fuzzilli, so maybe we can drop that patch entirely? In that case, please also update the [README](https://github.com/googleprojectzero/fuzzilli/blob/main/Targets/JavaScriptCore/README.md)

File: Sources/REPRLRun/main.swift [hunk 5]
```diff
@@ -37,7 +37,12 @@ if ctx == nil {
 }
 
 let argv = convertToCArray(Array(CommandLine.arguments[1...]))
+#if os(macOS)
+// Set the DYLD_FRAMEWORK_PATH env variable for the shell to run
```
Comment (saelo) [inline Sources/REPRLRun/main.swift:41]: This also feels misplaced here. Maybe we could just forward our environment to the child instead? Then you'd have to `DYLD_FRAMEWORK_PATH=xyz ./REPRLRun path/to/jsc` which would be ok I guess

File: Targets/JavaScriptCore/fuzzbuild.sh [hunk 6]
```diff
@@ -18,6 +18,10 @@ export WEBKIT_OUTPUTDIR=FuzzBuild
...
+    unset WEBKIT_OUTPUTDIR # remove the outputdir env on macOS for the build without cmake will be in the Source directory(maybe the build-jsc have problem)
```
Comment (saelo) [inline Targets/JavaScriptCore/fuzzbuild.sh:22]: I'm not sure I understand this. So if that environment variable exist, the build artifacts will be placed in Source/? Or the other way around?

**Review by saelo — COMMENTED** [review]
Thanks! A couple of questions and suggestions

File: Sources/Fuzzilli/Execution/REPRL.swift (hunk 1 above)
Comment (Eziodeng) [reply Sources/Fuzzilli/Execution/REPRL.swift:52]: Yes, truly it should .
(With the next problem, I find

File: Sources/Fuzzilli/Modules/NetworkSync.swift (hunk 2 above)
Comment (Eziodeng) [reply Sources/Fuzzilli/Modules/NetworkSync.swift:51]: Either old or new version of macOS with the build of the swift just have encountered the error with bellow 
```shell
fuzzilli/Sources/libsocket/include/libsocket.h:39:13: note: candidate found by name lookup is 'socket_t'
typedef int socket_t;
            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/sys/kernel_types.h:96:41: note: candidate found by name lookup is 'socket_t'
typedef struct __socket*                socket_t;
```
Maybe there is no problem on Linux. So I suggest avoid the `reference to ' socket_t' is ambiguous` error

File: Sources/FuzzilliCli/Profiles/JSCProfile.swift (hunk 3 above)
Comment (Eziodeng) [reply Sources/FuzzilliCli/Profiles/JSCProfile.swift:52]: I guess this just do not support for some old version, so just take care.
Here with the comment of `// Enable bounds check elimination validation`, I think there should be some function for that? So I donot handle with that.
Or just check the macOS sdk version to validate if this can be used for the build version of jsc? (For purpose of the old version? maybe this is not necessary)

File: Targets/JavaScriptCore/fuzzbuild.sh (hunk 4 above)
Comment (Eziodeng) [reply Targets/JavaScriptCore/fuzzbuild.sh:21]: I dont find that until you tell me. I just follow with the old one.   XD 
Yes we should just update the README for the workflow of the build and use for fuzzilli

But for the settings on Darwin, I think we still need to keep this shell with some args be set by `set-webkit-configuration`

File: Sources/REPRLRun/main.swift (hunk 5 above)
Comment (Eziodeng) [reply Sources/REPRLRun/main.swift:41]: Yes it should be fine, but I just find that with the static build of JavascriptCore of libJavascriptCore.a, we now do not need the DYLD_FRAMEWORK_PATH for the shell to run, so we can discard this commit at all if we properly solve the static build (Indeed a mess especially for some version of WebKit code fro some code between release tag). 

Also for the first problem

File: Targets/JavaScriptCore/fuzzbuild.sh (hunk 6 above)
Comment (Eziodeng) [reply Targets/JavaScriptCore/fuzzbuild.sh:22]: In the past, I have just set the sub project to build with xcode settings that reletive to source folder directly within opening the JavascriptCore proj

With the debug of my env, maybe my wrong settings for the sub xcode project make the problem, I will try with that again and check what the workflow of build with the perl script and Makefile.

Try with the export WEBKIT_OUTPUTDIR to FuzzBuild.
And the build script just lead to this pic which means if this env exist, the artifacts will just in the `Source/` folder
I think this should be the problem caused by the Makefile
image

**Comment by Eziodeng** [comment]
I now need some time to check if there is some difference between the static build on macOS  and the Linux one

**Comment by Eziodeng** [comment]
Working for other things now and take it off until free for sometime
